src/Npgsql/Replication/ReplicationConnection.cs

```diff
@@ -428,74 +421,100 @@ async Task<TimelineHistoryFile> TimelineHistoryInternal()
                 throw connector.UnexpectedMessageReceived(msg.Code);
             }
 
-            _sendFeedbackTimer.Change(WalReceiverStatusInterval, Timeout.InfiniteTimeSpan);
-            _requestFeedbackTimer.Change(_requestFeedbackInterval, Timeout.InfiniteTimeSpan);
-
             var buf = connector.ReadBuffer;
             var columnStream = new NpgsqlReadBuffer.ColumnStream(buf);
 
             SetTimeouts(_walReceiverTimeout, CommandTimeout);
-            using var _ = Defer(() => SetTimeouts(CommandTimeout, CommandTimeout));
-
-            while (true)
-            {
-                msg = await connector.ReadMessage(true, CancellationToken.None);
-                Expect<CopyDataMessage>(msg, connector);
 
-                // We received some message so there's no need to forcibly request feedback or time out.
-                // Reset the timer to request feedback.
-                _requestFeedbackTimer.Change(_requestFeedbackInterval, Timeout.InfiniteTimeSpan);
+            _sendFeedbackTimer = new Timer(TimerSendFeedback, state: null, WalReceiverStatusInterval, Timeout.InfiniteTimeSpan);
+            _requestFeedbackTimer = new Timer(TimerRequestFeedback, state: null, _requestFeedbackInterval, Timeout.InfiniteTimeSpan);
 
-                var messageLength = ((CopyDataMessage)msg).Length;
-                await buf.Ensure(1, true, CancellationToken.None);
-                var code = (char)buf.ReadByte();
-                switch (code)
-                {
-                case 'w': // XLogData
+            try
+            {
+                while (true)
                 {
-                    await buf.Ensure(24, true, CancellationToken.None);
-                    var startLsn = buf.ReadUInt64();
-                    var endLsn = buf.ReadUInt64();
-                    var sendTime = TimestampHandler.FromPostgresTimestamp(buf.ReadInt64()).ToLocalTime();
-
-                    if (unchecked((ulong)Interlocked.Read(ref _lastReceivedLsn)) < startLsn)
-                        Interlocked.Exchange(ref _lastReceivedLsn, unchecked((long)startLsn));
-                    if (unchecked((ulong)Interlocked.Read(ref _lastReceivedLsn)) < endLsn)
-                        Interlocked.Exchange(ref _lastReceivedLsn, unchecked((long)endLsn));
+                    msg = await connector.ReadMessage(true, CancellationToken.None);
+                    Expect<CopyDataMessage>(msg, connector);
 
-                    // dataLen = msg.Length - (code = 1 + walStart = 8 + walEnd = 8 + serverClock = 8)
-                    var dataLen = messageLength - 25;
-                    columnStream.Init(dataLen, canSeek: false);
-
-                    _cachedXLogDataMessage.Populate(new NpgsqlLogSequenceNumber(startLsn), new NpgsqlLogSequenceNumber(endLsn), sendTime, columnStream);
-                    yield return _cachedXLogDataMessage;
+                    // We received some message so there's no need to forcibly request feedback or time out.
```

Comment:
Since I moved the cancellation timeout to the buffers this comment is somewhat outdated.

```suggestion
                    // We received some message so there's no need to forcibly request feedback
```